### ISSUE TITLE
cannot build coredns with go 1.10

### ISSUE THREAD

**Issue by bgabler** [coredns/coredns#1540]
Trying to build coredns with a plugin and getting the following error:

../../go/src/github.com/coredns/coredns/core/dnsserver/register.go:33:13: cannot use newContext (type func() caddy.Context) as type func(*caddy.Instance) caddy.Context in field value

Seem to get this error when using `go get github.com/coredns/coredns` as well.

$ go version
go version go1.10 darwin/amd64

**Comment by fturib** [comment]
Hi @bgabler, I think you have not the correct version for the library caddy.
=> are you building coredns with the Makefile that is associated ?

Let me know.

**Comment by bgabler** [comment]
@fturib I've tried using `make` with the Makefile in the coredns repo.

I'm also getting that error when trying to `go get github.com/coredns/coredns`

```
[~/go/src/github.com]# go get github.com/coredns/coredns
# github.com/coredns/coredns/core/dnsserver
coredns/coredns/core/dnsserver/register.go:33:13: cannot use newContext (type func() caddy.Context) as type func(*caddy.Instance) caddy.Context in field value
```

**Comment by fturib** [comment]
Somehow that is linked to the version of github.com/mholt/caddy  you are building with.
I noticed that the signature of this function has changed lately (since one week) in caddy project.
However, coredns is expected to use a released version v0.10.10

(see in the Makefile).

Can it happen that you have have a vendor project inside your plugins's project that would contain a last version of caddy ?

**Comment by bgabler** [comment]
I tried manually putting the v10.10 source in place and still getting the same issue.

https://github.com/mholt/caddy/releases/download/v0.10.10/caddy_v0.10.10_linux_amd64.tar.gz

```
[~/go/src/github.com/coredns/coredns]# mv /ben/go/src/github.com/mholt/caddy-0.10.10 /ben/go/src/github.com/mholt/caddy
```
```
[~/go/src/github.com/coredns/coredns]# go build
# github.com/coredns/coredns/core/dnsserver
core/dnsserver/register.go:33:13: cannot use newContext (type func() caddy.Context) as type func(*caddy.Instance) caddy.Context in field value
```

When I remove everything related to caddy and run `make clean` then `make` it actually built this time. However, running `go build` will not work. 

¯\_(ツ)_/¯

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] cannot build ..." ]
>Hi @bgabler, I think you have not the correct version for the library caddy.
>=> are you building coredns with the Makefile that is associated ?
>
>Let me know.

Caddy made a backwards incompatible change in their master branch. We pin to the 
1.10.10 (or something) release in the Makefile. I'm waiting for Caddy to release 
so I can properly update CoreDNS (it's a 1 line change).

**Comment by Luzifer** [comment]
Maybe those pinnings from the `Makefile` should go into `Gopkg.toml` / `Gopkg.lock` and being vendored in `vendor`. That way a proper pinning would be in place without breaking `go get / build / install`.

**Comment by miekg** [comment]
#1523

**Comment by miekg** [comment]
Here's an idea (not sure how feasible). If we vendor everything and type alias things? Would that work?
Ugly and would still be whack-a-mole I'm afraid

**Comment by miekg** [comment]
This specific issue is fixed with #1544
